specifyweb/specify/tree_views.py

```diff
@@ -343,4 +351,392 @@ def perm_target(tree):
+    'SUBPHYLUM': 40,
+    'SUPERCLASS': 50,
+    'CLASS': 60,
+    'SUBCLASS': 70,
+    'INFRACLASS': 80,
+    'SUPERORDER': 90,
+    'ORDER': 100,
+    'SUBORDER': 110,
+    'INFRAORDER': 120,
+    'PARVORDER': 125,
+    'SUPERFAMILY': 130,
+    'FAMILY': 140,
+    'SUBFAMILY': 150,
+    'TRIBE': 160,
+    'SUBTRIBE': 170,
+    'GENUS': 180,
+    'SUBGENUS': 190,
+    'SECTION': 200,
+    'SUBSECTION': 210,
+    'SPECIES': 220,
+    'SUBSPECIES': 230,
+    'VARIETY': 240,
+    'SUBVARIETY': 250,
+    'FORMA': 260,
+    'SUBFORMA': 270
+}
+GEOGRAPHY_RANKS = {
+    'CONTINENT': 100,
+    'COUNTRY': 200,
+    'STATE': 300,
+    'COUNTY': 400
+}
+STORAGE_RANKS = {
+    'BUILDING': 100,
+    'COLLECTION': 150,
+    'ROOM': 200,
+    'AISLE': 250,
+    'CABINET': 300,
+    'SHELF': 350,
+    'BOX': 400,
+    'RACK': 450,
+    'VIAL': 500
+}
+GEOLOGIC_TIME_PERIOD_RANKS = {
+    'ERA': 100,
+    'PERIOD': 200,
+    'EPOCH': 300,
+    'AGE': 400
+}
+LITHO_STRAT_RANKS = {
+    'SUPERGROUP': 100,
+    'GROUP': 200,
+    'FORMATION': 300,
+    'MEMBER': 400,
+    'BED': 500
+}
+
+DEFAULT_RANK_INCREMENT = 100
+TAXON_RANK_INCREMENT = 10
+GEOGRAPHY_RANK_INCREMENT = DEFAULT_RANK_INCREMENT
+STORAGE_RANK_INCREMENT = 50
+GEOLOGIC_TIME_PERIOD_INCREMENT = DEFAULT_RANK_INCREMENT
+LITHO_STRAT_INCREMENT = DEFAULT_RANK_INCREMENT
+
+# Map tree type to default tree ranks and default rank id increment
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+# Wrapper for tree editing api functions
+def tree_edit(edit_func):
+    @login_maybe_required
+    @require_POST
+    @transaction.atomic
+    @wraps(edit_func)
+    def wrapper(*args, **kwargs):
+        try:
+            edit_func(*args, **kwargs)
+            result = {'success': True}
+        except BusinessRuleException as e:
+            result = {'success': False, 'error': str(e)}
+            return HttpResponseServerError(toJson(result), content_type="application/json")
+        return HttpResponse(toJson(result), content_type="application/json")
+
+    return wrapper
+
+@openapi(schema={
+    'post': {
+        "requestBody": {
+            "required": True,
+            "description": "Add rank to a tree.",
+            "content": {
+                "application/json": {
+                    "schema": {
+                        "type": "object",
+                        "description": "The request body.",
+                        "properties": {
+                            "newRankName": {
+                                "type": "string",
+                                "description": "The name of the new rank to add."
+                            },
+                            "parentRankName": {
```

Comment:
(optional) use parentRankId instead?

Reply:
how do you get the treeId in the frontend?

Reply:
two ways:
- the parent of the `Tree` component (TreeView) has a `treeDefinition` prop - from it, you can get tree id using `treeDefinition.id` (so need to pass that down as a prop to `Tree` component
- the tree component has an array from treeDefinitionItems. From there, you can get tree id by getting the id from the treeDef relationship: `const treeId = strictIdFromUrl(treeDefinitionItems[0].treeDef);`

Reply:
thank you :)